### PR TITLE
8342995: Enhance Attach API to support arbitrary length arguments - Linux

### PR DESCRIPTION
**Pull request by alexmenkov**
The fix updates Linux (and server-side of macosx) implementation to support Attach API v2 (shared code and Windows implementation were introduced by #20782)

Testing: tier1,tier2,tier3,tier4,hs-tier5-svc
manually tested backward compatibility (old tools can attach to current VMs, current tools can attach to older VMs) with jdk21 and jdk8.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8342995](https://bugs.openjdk.org/browse/JDK-8342995): Enhance Attach API to support arbitrary length arguments - Linux (**Sub-task** - P3)

### Reviewers
 * [Kevin Walls](https://openjdk.org/census#kevinw) (@kevinjwalls - **Reviewer**) Review applies to [2ca15a18](https://git.openjdk.org/jdk/pull/22223/files/2ca15a1894222034fac37a560f74b2440d34657f)
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22223/head:pull/22223` \
`$ git checkout pull/22223`

Update a local copy of the PR: \
`$ git checkout pull/22223` \
`$ git pull https://git.openjdk.org/jdk.git pull/22223/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22223`

View PR using the GUI difftool: \
`$ git pr show -t 22223`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22223.diff">https://git.openjdk.org/jdk/pull/22223.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22223#issuecomment-2484430876)
</details>

### CODE REVIEW

**Review by kevinjwalls — COMMENTED** [review]
Looks good I think.  (These are kind of hard to read, or hard to appreciate as a whole!)


VirtualMachineImpl.java and attachListener_PLATFORM.cpp have much in common between platforms, but maybe creating new common base classes is an interesting refactor work with not that much payoff.  Plus there are AIX versions that some of us don't build and test (which I think you mentioned elsewhere).



On Windows, attachListener_windows.cpp operations have:  assert(opened(), "must be");
...I was wondering if we don't need a similar assert in in src/hotspot/os/posix/attachListener_posix.cpp
But probably the different usage pattern explains that, i.e.
Windows' PipeChannel needs the open() method specifically called, but for Posix it creates the socket on construction.

(so this is fine, just some observations)

**Comment by alexmenkov** [comment]
> VirtualMachineImpl.java and attachListener_PLATFORM.cpp have much in common between platforms, but maybe creating new common base classes is an interesting refactor work with not that much payoff.

Yes, there a lot of common code between platforms and it can be refactored, but it would make the fix much bigger and harder to review.

> Plus there are AIX versions that some of us don't build and test (which I think you mentioned elsewhere).

Right. Most of the AIX code is copy of the posix/linux/macosx implementation, but I have no environment to build/test it, so I changed shared code to not break AIX build.

> On Windows, attachListener_windows.cpp operations have: assert(opened(), "must be"); ...I was wondering if we don't need a similar assert in in src/hotspot/os/posix/attachListener_posix.cpp But probably the different usage pattern explains that, i.e. Windows' PipeChannel needs the open() method specifically called, but for Posix it creates the socket on construction.

Right. The assert can be added to posix implementation, but it doesn't make much sense as SocketChannel does not opens the socket, but get it from the caller

**Review by kevinjwalls — APPROVED** [review]
Looks good, thanks, a couple of other notes here in case it helps anyone.


AttachOperation::write_reply(ReplyWriter * writer, jint result, const char* message, int message_len) 
is this only called by the write_reply method that passes message_len as strlen(stream) ?
The new check for len<0 might not really be useful but does no harm.

The existing "char msg[32]" is a buffer used to write the numeric jint result, not the message as the name might suggest.

HotSpotVirtualMachine.java already has writeCommand() that handles VERSION_2, so the Linux VirtualMachineImpl can just call it.  Also writeString is already there in HotSpotVirtualMachine, so removal from Linux is a good cleanup.


I'm sure we all want to find time for the future conversation about actually streaming the output rather than passing a single buffer around. 8-)

**Comment by alexmenkov** [comment]
> Looks good, thanks, a couple of other notes here in case it helps anyone.
> 
> AttachOperation::write_reply(ReplyWriter * writer, jint result, const char* message, int message_len) is this only called by the write_reply method that passes message_len as strlen(stream) ? The new check for len<0 might not really be useful but does no harm.

There are couple calls from `AttachOperation::read_request` that passes `const char*` (`message_len` is -1 as the default value).
I.e we need 2 versions of write_reply - one for `bufferedStream` and another for `const char*`.

> The existing "char msg[32]" is a buffer used to write the numeric jint result, not the message as the name might suggest.

This is old code (similar to deleted from `PosixAttachOperation::complete`).
Do you want to rename `msg` to `buf`?

> I'm sure we all want to find time for the future conversation about actually streaming the output rather than passing a single buffer around. 8-)

I'm open for it :)

**Comment by kevinjwalls** [comment]
> Do you want to rename msg to buf?
Don't worry about going back and renaming it just for me 8-)

File: src/hotspot/share/services/attachListener.hpp [hunk 1]
```diff
@@ -239,11 +245,16 @@ class AttachOperation: public CHeapObj<mtServiceability> {
     virtual int read(void* buffer, int size) = 0;
 
     // Reads unsigned value, returns -1 on error.
-    int read_uint();
+    //
+    // Attach client can make sanity connect/disconnect.
+    // In the case we get "premature EOF" error.
+    // If "may_be_empty" is true, the error is not logged.
```
Comment (sspitsyn) [inline src/hotspot/share/services/attachListener.hpp:251]: Nits or two minor suggestions:
 - replace: "In the case" => "In that case"
 - replace: "may_be_empty" => may_be_empty

File: src/hotspot/os/posix/attachListener_posix.cpp [hunk 2]
```diff
@@ -107,25 +97,61 @@ class PosixAttachListener: AllStatic {
...
+  void flush() override {
+    ::shutdown(_socket, SHUT_RDWR);
+  }
+};
+
 class PosixAttachOperation: public AttachOperation {
  private:
   // the connection to the client
-  int _socket;
+  SocketChannel _socket;
```
Comment (sspitsyn) [inline src/hotspot/os/posix/attachListener_posix.cpp:145]: Nit: The name `_socket` is misleading. I'd suggest to rename it to `_socket_channel`.

File: src/hotspot/share/services/attachListener.hpp (hunk 1 above)
Comment (alexmenkov) [reply src/hotspot/share/services/attachListener.hpp]: Fixed

File: src/hotspot/os/posix/attachListener_posix.cpp (hunk 2 above)
Comment (alexmenkov) [reply src/hotspot/os/posix/attachListener_posix.cpp]: Fixed

**Comment by alexmenkov** [comment]
> > Do you want to rename msg to buf?
> Don't worry about going back and renaming it just for me 8-)

Did it with the last update

**Review by sspitsyn — APPROVED** [review]
The fix looks good modulo a couple of nits.

File: src/hotspot/share/services/attachListener.cpp [hunk 3]
```diff
@@ -652,10 +658,26 @@ bool AttachOperation::read_request(RequestReader* reader) {
...
     return false;
   }
 
-  return read_request_data(reader, buffer_size, min_str_count, min_read_size);
+  bool result = read_request_data(reader, buffer_size, min_str_count, min_read_size);
+  if (result && ver == ATTACH_API_V1) {
+    // We know the whole request does not exceed buffer_size,
+    // for v1 also name/arguments should not exceed name_length_max/arg_length_max.
+    if (strlen(name()) > AttachOperation::name_length_max) {
+      log_error(attach)("Failed to read request: name is too long");
```
[the remark this answers is not in the document]
Comment (alexmenkov) [reply src/hotspot/share/services/attachListener.cpp]: There is "attach" in the log prefix, so I made it "Failed to read request: operation name is too long"
And updated similar logging for arguments several lines below

File: src/hotspot/share/services/attachListener.hpp [hunk 4]
```diff
@@ -231,6 +235,8 @@ class AttachOperation: public CHeapObj<mtServiceability> {
   // complete operation by sending result code and any result data to the client
   virtual void complete(jint result, bufferedStream* result_stream) = 0;
 
+  class ReplyWriter; //forward declaration
```
[the remark this answers is not in the document]
Comment (alexmenkov) [reply src/hotspot/share/services/attachListener.hpp]: Fixed

**Review by sspitsyn — APPROVED** [review]

**Comment by alexmenkov** [comment]
Kevin, Serguei, thank you for the review
/integrate

**Pull request closed without merging**